File: ex/hexagonal/src/main/kotlin/com/ktb/shop/application/ListProductsByCategory.kt
package com.ktb.shop.application

import com.ktb.shop.domain.CategoryId
import com.ktb.shop.domain.Product
import com.ktb.shop.domain.port.ProductCatalog

class ListProductsByCategory(private val catalog: ProductCatalog) {
    fun handle(categoryId: String): List<Product> =
        catalog.byCategory(CategoryId(categoryId))
}

File: ex/hexagonal-go/internal/application/list_products_by_category.go
package application

import (
	"context"

	"github.com/ktb/hexagonal-shop/internal/domain"
	"github.com/ktb/hexagonal-shop/internal/domain/port"
)

type ListProductsByCategory struct {
	catalog port.ProductCatalog
}

func NewListProductsByCategory(catalog port.ProductCatalog) *ListProductsByCategory {
	return &ListProductsByCategory{catalog: catalog}
}

func (uc *ListProductsByCategory) Handle(ctx context.Context, categoryID string) ([]domain.Product, error) {
	return uc.catalog.ByCategory(ctx, domain.NewCategoryID(categoryID))
}
